servicedef/sdk_config.go

```diff
@@ -1,50 +1,51 @@
 package servicedef
 
 import (
+	o "github.com/launchdarkly/sdk-test-harness/framework/opt"
+
 	"gopkg.in/launchdarkly/go-sdk-common.v2/ldtime"
 	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
-	"gopkg.in/launchdarkly/go-sdk-common.v2/ldvalue"
 )
 
 type SDKConfigParams struct {
-	Credential          string                              `json:"credential"`
-	StartWaitTimeMS     ldtime.UnixMillisecondTime          `json:"startWaitTimeMs,omitempty"`
-	InitCanFail         bool                                `json:"initCanFail,omitempty"`
-	Streaming           *SDKConfigStreamingParams           `json:"streaming,omitempty"`
-	Events              *SDKConfigEventParams               `json:"events,omitempty"`
-	PersistentDataStore *SDKConfigPersistentDataStoreParams `json:"persistentDataStore,omitempty"`
-	BigSegments         *SDKConfigBigSegmentsParams         `json:"bigSegments,omitempty"`
-	Tags                *SDKConfigTagsParams                `json:"tags,omitempty"`
```

Comment:
The main advantages of changing these things from pointers to Maybes are 1. it prevents accidental nil panics, and 2. it requires us to use a safe "copy, mutate, copy back" pattern if we want to modify fields within them. The secondary consequence, from the test _service_'s point of view, is that there will be a bunch of explicit nulls if these are unset, instead of just omitting the property, because `json.Marshal`'s `omitempty` behavior _still_ doesn't support dropping zero-valued things that are not of simple types. I left the `omitempty` tags in here just to convey what we would prefer the behavior to be. Maybe one of these days Go will improve it.

Reply:
Btw, not surprisingly, the `omitempty` for custom types issue has been raised many times over the years with various reasonable proposals on how to address it, and the Go maintainers have more or less blown off every proposal like "eh, we don't think this is necessary." But their attitude may have changed in the last few years, who knows. I left a comment [here](https://github.com/golang/go/issues/11939#issuecomment-1098521910) about the particular significance of this with generics.